File: java/src/main/java/com/bakobo/fiki/Sfv.java
package com.bakobo.fiki;

import java.util.ArrayList;
import java.util.Base64;
import java.util.List;
import java.util.Map;

/**
 * The RFC 8941 subset RFC 9421 actually uses (this.i @2q9gv70t, @2tt6fmc0).
 *
 * <p>Hand-rolled rather than depended upon, for the reason every port hand-rolls it: fiki's slice
 * of structured fields is small and CLOSED — a dictionary whose members are inner lists of strings
 * with parameters, plus byte sequences — and the shared vectors pin its entire output surface byte
 * for byte. The usual argument against writing your own parser holds where the grammar is
 * open-ended; this one's every output is checked against committed bytes shared with four other
 * implementations.
 *
 * <p>What is deliberately NOT here: decimals, tokens, inner-list items with their own parameters,
 * and every field type RFC 9421 never puts in these two headers.
 */
final class Sfv {

    private Sfv() {}

    /**
     * This class's own failure, which never escapes the package: the parser cannot know WHICH
     * header it is reading, and the taxonomy distinguishes an unparsable Signature from an
     * unparsable Signature-Input, so callers translate it into the kind that names the header.
     */
    static final class SyntaxException extends RuntimeException {
        SyntaxException(String message) {
            super(message);
        }
    }

    /** A covered-component list with its parameters, in the order they arrived. */
    record InnerList(List<String> items, List<Map.Entry<String, Object>> params) {
        Object param(String key) {
            for (Map.Entry<String, Object> entry : params) {
                if (entry.getKey().equals(key)) {
                    return entry.getValue();
                }
            }
            return null;
        }
    }

    /** A dictionary member: either an inner list, or a bare value with parameters. */
    record Member(String key, InnerList list, Object value) {}

    private static final class Cursor {
        private final String text;
        private int at;

        Cursor(String text) {
            this.text = text;
        }

        boolean done() {
            return at >= text.length();
        }

        char peek() {
            return done() ? '\0' : text.charAt(at);
        }

        void skipSpace() {
            while (!done() && (peek() == ' ' || peek() == '\t')) {
                at++;
            }
        }

        void expect(char ch) {
            if (done() || peek() != ch) {
                throw new SyntaxException("expected " + ch + " at offset " + at);
            }
            at++;
        }

        String parseKey() {
            if (done() || !(isLower(peek()) || peek() == '*')) {
                throw new SyntaxException("a key starts with a lowercase letter or *");
            }
            int start = at;
            while (!done() && (isLower(peek()) || isDigit(peek()) || "_-.*".indexOf(peek()) >= 0)) {
                at++;
            }
            return text.substring(start, at);
        }

        String parseString() {
            at++; // the opening quote, which the caller already peeked
            StringBuilder out = new StringBuilder();
            while (!done()) {
                char ch = text.charAt(at++);
                if (ch == '\\') {
                    if (done()) {
                        throw new SyntaxException("a string ended mid-escape");
                    }
                    char escaped = text.charAt(at++);
                    if (escaped != '"' && escaped != '\\') {
                        throw new SyntaxException("only \\\" and \\\\ may be escaped");
                    }
                    out.append(escaped);
                } else if (ch == '"') {
                    return out.toString();
                } else {
                    out.append(ch);
                }
            }
            throw new SyntaxException("a string ran to the end of the field");
        }

        byte[] parseByteSequence() {
            at++; // the opening colon
            int start = at;
            while (!done() && peek() != ':') {
                at++;
            }
            String encoded = text.substring(start, at);
            expect(':');
            try {
                return Base64.getDecoder().decode(encoded);
            } catch (IllegalArgumentException e) {
                throw new SyntaxException("a byte sequence must be base64 between colons");
            }
        }

        long parseInteger() {
            int start = at;
            if (peek() == '-') {
                at++;
            }
            while (!done() && isDigit(peek())) {
                at++;
            }
            try {
                return Long.parseLong(text.substring(start, at));
            } catch (NumberFormatException e) {
                throw new SyntaxException("expected an integer at offset " + start);
            }
        }

        Object parseBareItem() {
            char ch = peek();
            if (ch == '"') {
                return parseString();
            }
            if (ch == ':') {
                return parseByteSequence();
            }
            if (ch == '?') {
                at++;
                if (done()) {
                    throw new SyntaxException("a boolean is ?0 or ?1");
                }
                char flag = text.charAt(at++);
                if (flag != '0' && flag != '1') {
                    throw new SyntaxException("a boolean is ?0 or ?1");
                }
                return flag == '1';
            }
            if (ch == '-' || isDigit(ch)) {
                return parseInteger();
            }
            throw new SyntaxException("unsupported item at offset " + at);
        }

        List<Map.Entry<String, Object>> parseParameters() {
            List<Map.Entry<String, Object>> params = new ArrayList<>();
            while (!done() && peek() == ';') {
                at++;
                skipSpace();
                String key = parseKey();
                if (!done() && peek() == '=') {
                    at++;
                    params.add(Map.entry(key, parseBareItem()));
                } else {
                    params.add(Map.entry(key, Boolean.TRUE));
                }
            }
            return params;
        }

        InnerList parseInnerList() {
            at++; // the opening parenthesis
            List<String> items = new ArrayList<>();
            while (true) {
                skipSpace();
                if (done()) {
                    throw new SyntaxException("an inner list ran to the end of the field");
                }
                if (peek() == ')') {
                    at++;
                    break;
                }
                Object item = parseBareItem();
                if (!(item instanceof String text)) {
                    throw new SyntaxException("fiki's covered components are strings");
                }
                // RFC 9421 never puts parameters on the members of a covered-component list.
                if (!done() && peek() == ';') {
                    throw new SyntaxException("parameters on a covered component");
                }
                if (!done() && peek() != ' ' && peek() != ')') {
                    throw new SyntaxException("expected a space or ) at offset " + at);
                }
                items.add(text);
            }
            return new InnerList(items, parseParameters());
        }
    }

    private static boolean isLower(char ch) {
        return ch >= 'a' && ch <= 'z';
    }

    private static boolean isDigit(char ch) {
        return ch >= '0' && ch <= '9';
    }

    /** Parse an RFC 8941 dictionary, preserving member order because the verify side needs it. */
    static List<Member> parseDictionary(String text) {
        Cursor cursor = new Cursor(text);
        List<Member> out = new ArrayList<>();
        cursor.skipSpace();
        while (!cursor.done()) {
            String key = cursor.parseKey();
            Member member;
            if (!cursor.done() && cursor.peek() == '=') {
                cursor.at++;
                if (cursor.peek() == '(') {
                    member = new Member(key, cursor.parseInnerList(), null);
                } else {
                    Object value = cursor.parseBareItem();
                    member = new Member(key, new InnerList(List.of(), cursor.parseParameters()), value);
                }
            } else {
                member = new Member(key, new InnerList(List.of(), cursor.parseParameters()), Boolean.TRUE);
            }
            out.removeIf(existing -> existing.key().equals(key));
            out.add(member);
            cursor.skipSpace();
            if (cursor.done()) {
                break;
            }
            cursor.expect(',');
            cursor.skipSpace();
            if (cursor.done()) {
                throw new SyntaxException("a dictionary ended with a trailing comma");
            }
        }
        return out;
    }

    static String serializeBareItem(Object value) {
        if (value instanceof String text) {
            return '"' + text.replace("\\", "\\\\").replace("\"", "\\\"") + '"';
        }
        if (value instanceof Long n) {
            return n.toString();
        }
        if (value instanceof byte[] raw) {
            return ':' + Base64.getEncoder().encodeToString(raw) + ':';
        }
        // Only FALSE reaches here: RFC 8941 renders a true-valued parameter as a bare key, which
        // serializeParameters does before calling this, and fiki never puts a boolean in an item
        // position.
        return "?0";
    }

    static String serializeParameters(List<Map.Entry<String, Object>> params) {
        StringBuilder out = new StringBuilder();
        for (Map.Entry<String, Object> entry : params) {
            if (Boolean.TRUE.equals(entry.getValue())) {
                out.append(';').append(entry.getKey());
            } else {
                out.append(';').append(entry.getKey()).append('=').append(serializeBareItem(entry.getValue()));
            }
        }
        return out.toString();
    }

    /** Render a covered-component list with its signature parameters. */
    static String serializeInnerList(InnerList list) {
        StringBuilder out = new StringBuilder("(");
        for (int i = 0; i < list.items().size(); i++) {
            if (i > 0) {
                out.append(' ');
            }
            out.append(serializeBareItem(list.items().get(i)));
        }
        return out.append(')').append(serializeParameters(list.params())).toString();
    }
}

File: rust/src/sfv.rs
//! The RFC 8941 subset RFC 9421 actually uses (`this.i` @2q9gv70t, @2tt6fmc0).
//!
//! Hand-rolled rather than depended upon, for the reason every port hand-rolls it: fiki's slice of
//! structured fields is small and CLOSED — a dictionary whose members are inner lists of strings
//! with parameters, plus byte sequences — and the shared vectors pin its entire output surface byte
//! for byte. The usual argument against writing your own parser holds where the grammar is
//! open-ended; this one's every output is checked against committed bytes shared with four other
//! implementations.

use crate::keys::{b64std, b64std_decode};

/// A parameter value. The variants are exactly what RFC 9421 puts in these two headers.
#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) enum Value {
    Text(String),
    Integer(i64),
    Boolean(bool),
    Bytes(Vec<u8>),
}

/// A covered-component list with its signature parameters, in the order they arrived.
///
/// Order is load-bearing on the verify side: a verifier that reorders what it received computes a
/// different base and rejects a good signature.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub(crate) struct InnerList {
    pub items: Vec<String>,
    pub params: Vec<(String, Value)>,
}

impl InnerList {
    pub fn param(&self, key: &str) -> Option<&Value> {
        self.params.iter().find(|(k, _)| k == key).map(|(_, v)| v)
    }
}

#[derive(Debug, Clone)]
pub(crate) struct Member {
    pub list: InnerList,
    pub value: Option<Value>,
}

/// This module's own failure. It never escapes: the parser cannot know WHICH header it is reading,
/// and the taxonomy distinguishes an unparsable Signature from an unparsable Signature-Input, so
/// callers translate it into the kind that names the header.
#[derive(Debug)]
pub(crate) struct SyntaxError;

type Parsed<T> = std::result::Result<T, SyntaxError>;

struct Cursor<'a> {
    text: &'a [u8],
    at: usize,
}

impl<'a> Cursor<'a> {
    fn done(&self) -> bool {
        self.at >= self.text.len()
    }

    fn peek(&self) -> u8 {
        if self.done() {
            0
        } else {
            self.text[self.at]
        }
    }

    fn skip_space(&mut self) {
        while !self.done() && (self.peek() == b' ' || self.peek() == b'\t') {
            self.at += 1;
        }
    }

    fn expect(&mut self, ch: u8) -> Parsed<()> {
        if self.done() || self.peek() != ch {
            return Err(SyntaxError);
        }
        self.at += 1;
        Ok(())
    }

    fn parse_key(&mut self) -> Parsed<String> {
        let start = self.at;
        if self.done() || !(self.peek().is_ascii_lowercase() || self.peek() == b'*') {
            return Err(SyntaxError);
        }
        while !self.done()
            && (self.peek().is_ascii_lowercase()
                || self.peek().is_ascii_digit()
                || matches!(self.peek(), b'_' | b'-' | b'.' | b'*'))
        {
            self.at += 1;
        }
        String::from_utf8(self.text[start..self.at].to_vec()).map_err(|_| SyntaxError)
    }

    fn parse_string(&mut self) -> Parsed<String> {
        self.at += 1; // the opening quote, which the caller already peeked
        let mut out = String::new();
        while !self.done() {
            let ch = self.text[self.at];
            self.at += 1;
            match ch {
                b'\\' => {
                    if self.done() {
                        return Err(SyntaxError);
                    }
                    let esc = self.text[self.at];
                    self.at += 1;
                    if esc != b'"' && esc != b'\\' {
                        return Err(SyntaxError);
                    }
                    out.push(esc as char);
                }
                b'"' => return Ok(out),
                _ => out.push(ch as char),
            }
        }
        Err(SyntaxError)
    }

    fn parse_byte_sequence(&mut self) -> Parsed<Vec<u8>> {
        self.at += 1; // the opening colon
        let start = self.at;
        while !self.done() && self.peek() != b':' {
            self.at += 1;
        }
        let encoded = std::str::from_utf8(&self.text[start..self.at]).map_err(|_| SyntaxError)?;
        let encoded = encoded.to_owned();
        self.expect(b':')?;
        b64std_decode(&encoded).ok_or(SyntaxError)
    }

    fn parse_integer(&mut self) -> Parsed<i64> {
        let start = self.at;
        if self.peek() == b'-' {
            self.at += 1;
        }
        while !self.done() && self.peek().is_ascii_digit() {
            self.at += 1;
        }
        std::str::from_utf8(&self.text[start..self.at])
            .map_err(|_| SyntaxError)?
            .parse()
            .map_err(|_| SyntaxError)
    }

    fn parse_bare_item(&mut self) -> Parsed<Value> {
        match self.peek() {
            b'"' => Ok(Value::Text(self.parse_string()?)),
            b':' => Ok(Value::Bytes(self.parse_byte_sequence()?)),
            b'?' => {
                self.at += 1;
                if self.done() {
                    return Err(SyntaxError);
                }
                let flag = self.text[self.at];
                self.at += 1;
                match flag {
                    b'0' => Ok(Value::Boolean(false)),
                    b'1' => Ok(Value::Boolean(true)),
                    _ => Err(SyntaxError),
                }
            }
            ch if ch == b'-' || ch.is_ascii_digit() => Ok(Value::Integer(self.parse_integer()?)),
            _ => Err(SyntaxError),
        }
    }

    fn parse_parameters(&mut self) -> Parsed<Vec<(String, Value)>> {
        let mut params = Vec::new();
        while !self.done() && self.peek() == b';' {
            self.at += 1;
            self.skip_space();
            let key = self.parse_key()?;
            if !self.done() && self.peek() == b'=' {
                self.at += 1;
                params.push((key, self.parse_bare_item()?));
            } else {
                params.push((key, Value::Boolean(true)));
            }
        }
        Ok(params)
    }

    fn parse_inner_list(&mut self) -> Parsed<InnerList> {
        self.at += 1; // the opening parenthesis
        let mut items = Vec::new();
        loop {
            self.skip_space();
            if self.done() {
                return Err(SyntaxError);
            }
            if self.peek() == b')' {
                self.at += 1;
                break;
            }
            let item = match self.parse_bare_item()? {
                Value::Text(text) => text,
                // fiki's covered components are strings; anything else is a shape it cannot
                // render back.
                _ => return Err(SyntaxError),
            };
            // RFC 9421 never puts parameters on the members of a covered-component list.
            if !self.done() && self.peek() == b';' {
                return Err(SyntaxError);
            }
            if !self.done() && self.peek() != b' ' && self.peek() != b')' {
                return Err(SyntaxError);
            }
            items.push(item);
        }
        Ok(InnerList {
            items,
            params: self.parse_parameters()?,
        })
    }
}

/// Parse an RFC 8941 dictionary, preserving member order because the verify side depends on it.
pub(crate) fn parse_dictionary(text: &str) -> Parsed<Vec<(String, Member)>> {
    let mut cursor = Cursor {
        text: text.as_bytes(),
        at: 0,
    };
    let mut out: Vec<(String, Member)> = Vec::new();
    cursor.skip_space();
    while !cursor.done() {
        let key = cursor.parse_key()?;
        let member = if !cursor.done() && cursor.peek() == b'=' {
            cursor.at += 1;
            if cursor.peek() == b'(' {
                Member {
                    list: cursor.parse_inner_list()?,
                    value: None,
                }
            } else {
                let value = cursor.parse_bare_item()?;
                let params = cursor.parse_parameters()?;
                Member {
                    list: InnerList {
                        items: Vec::new(),
                        params,
                    },
                    value: Some(value),
                }
            }
        } else {
            let params = cursor.parse_parameters()?;
            Member {
                list: InnerList {
                    items: Vec::new(),
                    params,
                },
                value: Some(Value::Boolean(true)),
            }
        };
        out.retain(|(existing, _)| existing != &key);
        out.push((key, member));
        cursor.skip_space();
        if cursor.done() {
            break;
        }
        cursor.expect(b',')?;
        cursor.skip_space();
        if cursor.done() {
            return Err(SyntaxError);
        }
    }
    Ok(out)
}

fn serialize_bare_item(value: &Value) -> String {
    match value {
        Value::Text(text) => format!("\"{}\"", text.replace('\\', "\\\\").replace('"', "\\\"")),
        Value::Integer(n) => n.to_string(),
        Value::Bytes(raw) => format!(":{}:", b64std(raw)),
        // Only `false` reaches here: RFC 8941 renders a true-valued parameter as a bare key, which
        // serialize_parameters does before calling this, and fiki never puts a boolean in an item
        // position.
        Value::Boolean(_) => "?0".to_string(),
    }
}

fn serialize_parameters(params: &[(String, Value)]) -> String {
    params
        .iter()
        .map(|(key, value)| match value {
            Value::Boolean(true) => format!(";{key}"),
            _ => format!(";{key}={}", serialize_bare_item(value)),
        })
        .collect()
}

/// Render a covered-component list with its signature parameters.
pub(crate) fn serialize_inner_list(list: &InnerList) -> String {
    let items: Vec<String> = list
        .items
        .iter()
        .map(|item| serialize_bare_item(&Value::Text(item.clone())))
        .collect();
    format!(
        "({}){}",
        items.join(" "),
        serialize_parameters(&list.params)
    )
}
